### DISCUSSION TITLE
Adding key-code and scan-code fields to KeyEvent struct

### DISCUSSION THREAD

**Discussion by quendimax** [slint-ui/slint#3189] [Ideas] [answered]
Hi slint team!

Is there any plans to add API for scan/key-code into KeyEvent? I saw the discussion in [this issue](https://github.com/slint-ui/slint/issues/144), and if I understand right, the result is passing all keys via unicode in the `KeyEvent.text` field. But this solution doesn't cover all possible use cases. Mostly because the value of the `text` field depends on user's keyboard layout. So it is impossible to create a shortcut that would work for everyone (as Ctrl+V for example works everywhere independently on your system language/current keyboard layout).

I think adding `scancode` and `keycode` to `KeyEvent` would add much more flexibility for developer to implement his/her ideas.

Thanks

**Comment by ogoffart** [comment]
What usecase do you have in mind?

For example, you can check if the key `"v"` is pressed with the ctrl key, and it will be the V key on all layout.
In general, shortcuts don't depends on the user's keyboard layout. 
For example, the shortcut for Select all is always Ctrl+A (with the 'A' key) regardless where the A key is on the keyboard.

**Reply by quendimax** [reply]
> For example, you can check if the key "v" is pressed with the ctrl key, and it will be the V key on all layout.

Yes. All OS that I have tested generate English letters when they are combined with Ctrl/Cmd key independently on the current language layout. But if I use combinations with Alt/AltGr/Shift, I see the layout related symbols (as expected).

My use case is an app with keybindings a la vim, and I want to use a Key with and without Modifiers as a Keybinding. And I'd like to avoid the vim problem with non english layouts, when default keybindings work only for the standard English layout. For vim this problem has some workarounds, but no one resolves the issue entirely, and everyone requires high enough user qualification.

Also as an example, I use many Alt+Key shortcuts in Visual Studio and VSCode, and they work well (actually VSCode requires to set the "keyboard.dispatch" option to "keyCode" value on Linux and MacOS). So it doesn't look a rare use case.

**Reply by quendimax** [reply]
Also it looks like now it is impossible to distinguish numpad keys from usual ones.

**Reply by hunger** [reply, marked as answer]
The next winit version will have a greatly revamped key event system. We should be able to benefit from the extra information they provide and should be able to improve our key events accordingly. 🤞🏼
